### PR TITLE
bpo-39648:[WIP]updated math.gcd to accept 'n' arguments.

### PR DESCRIPTION
**Pull request by ananthan-123**
https://bugs.python.org/issue39648

### CODE REVIEW

**Comment by ananthan-123** [comment]
I made clinic and now it is showing checksum error.
What to do?

**Comment by mdickinson** [comment]
What error are you getting `make clinic` runs fine for me on commit 660448e646ae71049273499c1062badc68de2c16, and has no effect on `mathmodule.c.h`, so it looks as though everything is up to date here.

**Comment by mdickinson** [comment]
Ah, I think I understand: after you merge the upstream master in, you'll get conflicts in mathmodule.c.h. You can either restore mathmodule.c.h to either the local version or the upstream version (check out `git checkout --ours` and `git checkout --theirs`), or perhaps more simply, I believe it should always be safe to simply delete `mathmodule.c.h` entirely and then regenerate using `make clinic`.

**Comment by ananthan-123** [comment]
I think the test failure is due to some other problem!!

File: Modules/mathmodule.c [hunk 1]
```diff
@@ -824,37 +824,44 @@ m_log10(double x)
...
-    PyObject *g;
-
-    a = PyNumber_Index(a);
-    if (a == NULL)
-        return NULL;
-    b = PyNumber_Index(b);
-    if (b == NULL) {
-        Py_DECREF(a);
-        return NULL;
+    PyObject *g = 0, *item, *in;
```
Comment (sweeneyde) [inline Modules/mathmodule.c]:
```suggestion
    PyObject *g, tmp, *item, *in;
    Py_ssize_t i;
    
    g = _PyLong_Zero;
    Py_INCREF(_PyLong_Zero);
```

File: Modules/mathmodule.c [hunk 2]
```diff
@@ -824,37 +824,44 @@ m_log10(double x)
...
+    
+    Py_ssize_t i;
+    for (i = 0; i < n; i++) {
+        item = args[i];
+        in = PyNumber_Index(item);
+        if (in == NULL) {
+            return NULL;
+        }
+        g = _PyLong_GCD(g, in);
```
Comment (sweeneyde) [inline Modules/mathmodule.c:842]:
```suggestion
        tmp = _PyLong_GCD(g, in);
        Py_DECREF(g);
        g = tmp;
```

File: Lib/test/test_math.py [hunk 3]
```diff
@@ -691,20 +691,32 @@ def msum(iterable):
...
         self.assertEqual(gcd(0, -1), 1)
         self.assertEqual(gcd(7, 1), 1)
         self.assertEqual(gcd(7, -1), 1)
+        self.assertEqual(gcd(7, 1, 0 ,2 ,33, 44, 56, 67, 78, 89), 0)
         self.assertEqual(gcd(-23, 15), 1)
         self.assertEqual(gcd(120, 84), 12)
         self.assertEqual(gcd(84, -120), 12)
         self.assertEqual(gcd(1216342683557601535506311712,
-                             436522681849110124616458784), 32)
+                             436522681849110124616458784,
```
Comment (sweeneyde) [inline Lib/test/test_math.py:718]: Existing tests are sacred, so I think new tests should be independently added, not tweaked from existing tests.

File: Lib/test/test_math.py [hunk 4]
```diff
@@ -691,20 +691,32 @@ def msum(iterable):
             s = msum(vals)
             self.assertEqual(msum(vals), math.fsum(vals))
 
-    def testGcd(self):
+    def test_gcd(self):
         gcd = math.gcd
-        self.assertEqual(gcd(0, 0), 0)
+        self.assertEqual(gcd(), 0)
```
Comment (sweeneyde) [inline Lib/test/test_math.py:697]: Keep the `self.assertEqual(gcd(0, 0), 0)` test.

File: Doc/library/math.rst [hunk 5]
```diff
@@ -126,14 +126,21 @@ Number-theoretic and representation functions
...
 
 
-.. function:: gcd(a, b)
+.. function:: gcd(*arguments)
 
-   Return the greatest common divisor of the integers *a* and *b*.  If either
-   *a* or *b* is nonzero, then the value of ``gcd(a, b)`` is the largest
-   positive integer that divides both *a* and *b*.  ``gcd(0, 0)`` returns
-   ``0``.
+   returns the greatest common divisor of multiple scalar arguments.
```
Comment (sweeneyde) [inline Doc/library/math.rst]:
```suggestion
   Return the greatest common divisor of multiple integer arguments.
```

File: Doc/library/math.rst [hunk 6]
```diff
@@ -126,14 +126,21 @@ Number-theoretic and representation functions
...
+   For example,
```
Comment (sweeneyde) [inline Doc/library/math.rst]: I think the following precise statement (inspired by the existing documentation) with relevant examples is clearer than a more vague description followed by a list of "some exceptions."
```suggestion
    If any one of the *arguments* is nonzero, then the value of ``gcd(*arguments)`` is the largest positive integer that divides all of the *arguments*. If all of the *arguments* are zero, then ``gcd(*arguments)`` returns ``0``. For example::
   
        >>> gcd(22, 33)
        11
        >>> gcd(-6, 8, 10, 12)
        2
        >>> gcd(-5)
        5
        >>> gcd(0, 0)
        0
        >>> gcd()
        0

```

File: Modules/mathmodule.c [hunk 7]
```diff
@@ -824,37 +824,44 @@ m_log10(double x)
...
     }
-    g = _PyLong_GCD(a, b);
-    Py_DECREF(a);
-    Py_DECREF(b);
     return g;
 }
 
+PyDoc_STRVAR(math_gcd__doc,
+             "gcd(*arguments) -> value\n\n\
+returns the greatest common divisor of multiple scalar arguments.\n\
```
Comment (sweeneyde) [inline Modules/mathmodule.c]: (Same comment as in the docs)

**Comment by mdickinson** [comment]
> I think the test failure is due to some other problem!!

Yes. You need to initialise `g` to be the integer zero (it's currently initialised to the `NULL` pointer, and then that `NULL` pointer gets passed into the first call to `_PyLong_GCD`). After that, there are some reference leaks to sort out.

File: Lib/test/test_math.py [hunk 8]
```diff
@@ -691,20 +691,32 @@ def msum(iterable):
             s = msum(vals)
             self.assertEqual(msum(vals), math.fsum(vals))
 
-    def testGcd(self):
+    def test_gcd(self):
```
Comment (sweeneyde) [inline Lib/test/test_math.py:694]: I think leaving the name unchanged is a good idea for when someone wants to, say, bisect to see when some test case started failing/running slowly/memory leaking.

File: Lib/test/test_math.py [hunk 9]
```diff
@@ -691,20 +691,32 @@ def msum(iterable):
...
+        self.assertEqual(gcd(0, 2, 3, 4, 5, 6, 7, 8, 9), 0)
+        self.assertEqual(gcd(222, 444, 666, 888, 990, 224, 556, 888, 228,
+                             886, 8890, 440, 228, 884), 2)
         self.assertEqual(gcd(1, 0), 1)
         self.assertEqual(gcd(-1, 0), 1)
         self.assertEqual(gcd(0, 1), 1)
         self.assertEqual(gcd(0, -1), 1)
         self.assertEqual(gcd(7, 1), 1)
         self.assertEqual(gcd(7, -1), 1)
+        self.assertEqual(gcd(7, 1, 0 ,2 ,33, 44, 56, 67, 78, 89), 0)
```
Comment (sweeneyde) [inline Lib/test/test_math.py]:
```suggestion
        self.assertEqual(gcd(7, 1, 0 ,2 ,33, 44, 56, 67, 78, 89), 1)
```

File: Lib/test/test_math.py [hunk 10]
```diff
@@ -731,7 +743,7 @@ def testGcd(self):
         self.assertEqual(gcd(-b, -a), c)
 
         self.assertRaises(TypeError, gcd, 120.0, 84)
-        self.assertRaises(TypeError, gcd, 120, 84.0)
+        self.assertRaises(TypeError, gcd, 120.0, 84, 23.5, 34.0, 78.0)
```
Comment (sweeneyde) [inline Lib/test/test_math.py]: Existing tests are sacred, so I think new tests should be independently added, not tweaked from existing tests.

File: Modules/mathmodule.c [hunk 11]
```diff
@@ -824,37 +824,44 @@ m_log10(double x)
...
-    if (b == NULL) {
-        Py_DECREF(a);
-        return NULL;
+    PyObject *g = 0, *item, *in;
+    
+    Py_ssize_t i;
+    for (i = 0; i < n; i++) {
+        item = args[i];
+        in = PyNumber_Index(item);
+        if (in == NULL) {
```
Comment (sweeneyde) [inline Modules/mathmodule.c:838]:
```suggestion
        if (in == NULL) {
            Py_DECREF(g);
```

**Review by sweeneyde — COMMENTED** [review]
I think the main bug is that there is a difference between the C value `0` (aka a null pointer) and the Python `int` `0`.

**Comment by ananthan-123** [comment]
Updated the code.
Do we have to decref 'item'.

File: Modules/mathmodule.c [hunk 12]
```diff
@@ -824,37 +824,44 @@ m_log10(double x)
...
-
-    a = PyNumber_Index(a);
-    if (a == NULL)
-        return NULL;
-    b = PyNumber_Index(b);
-    if (b == NULL) {
-        Py_DECREF(a);
-        return NULL;
+    PyObject *g, *item, *in;
+    g = PyLong_FromLong(0);
```
Comment (mdickinson) [inline Modules/mathmodule.c:831]: All Python C-API calls need to have their return value checked. That includes this one, and the `_PyLong_GCD` call below.

File: Modules/mathmodule.c (hunk 2 above)
Comment (mdickinson) [reply Modules/mathmodule.c:842]: Using `Py_SETREF(g, _PyLong_GCD(g, in));` is another possibility here.

**Comment by mdickinson** [comment]
> Do we have to decref 'item'.

No: `item` is effectively a borrowed reference; no need to decref. Having said that, you don't really need a separate `item` at all. You could replace:
```
item = args[i];
in = PyNumber_Index(item);
```

with simply
```
in = PyNumber_Index(args[i]);
```

**Comment by ananthan-123** [comment]
Thanks for the review.Made suitable changes.

**Comment by ananthan-123** [comment]
Now the problem is with clinic(don't know if any other problem is there).Don't we have to make clinic.
Do we have to regain previous clinic file.

**Comment by mdickinson** [comment]
Closing, since #18604 was merged.

**Pull request closed without merging**